README.md
# Starting blocks
## A page transition and blocks ES6 framework by REZO ZERO

[![npm](https://img.shields.io/npm/l/starting-blocks.svg)](https://www.npmjs.com/package/starting-blocks)
[![npm](https://img.shields.io/npm/v/starting-blocks.svg)](https://www.npmjs.com/package/starting-blocks)
[![Build Status](https://travis-ci.org/rezozero/starting-blocks.svg?branch=master)](https://travis-ci.org/rezozero/starting-blocks)
![64kB size](https://img.shields.io/badge/Size-64kB-green.svg)

- [Adrien Scholaert](https://github.com/Gouterman)
- [Ambroise Maupate](https://github.com/ambroisemaupate)
- [Quentin Neyraud](https://github.com/quentinneyraud)
- [Maxime Bérard](https://github.com/maximeberard)

## Spec

- Compatible from **Internet Explorer 10+**
- *jQuery* 3.3.1
- *jquery.waitforimages* (for dispatching *onLoad* events to pages and blocks)
- *vanilla-lazyload* (for optional automatic image lazyloading)
- *debounce* (http://davidwalsh.name/javascript-debounce-function)
- *loglevel*
- Native *window.Promise*. Make sure to use a polyfill for [Internet Explorer 10 - 11](https://cdnjs.cloudflare.com/ajax/libs/es6-promise/4.1.0/es6-promise.auto.js)
- Native *window.MutationObserver*. Make sure to use a polyfill for [Internet Explorer 10](https://cdnjs.cloudflare.com/ajax/libs/MutationObserver.js/0.3.2/mutationobserver.min.js)
- Native *window.fetch* method to perform any XHR requests. If *window.fetch* is not available, XHR will be disabled. Make sure to use a polyfill for [Internet Explorer 10-11](https://cdnjs.cloudflare.com/ajax/libs/fetch/2.0.3/fetch.min.js)
- *Webpack* (for development)

```html
<script src="https://cdnjs.cloudflare.com/ajax/libs/es6-promise/4.1.1/es6-promise.auto.min.js"
        integrity="sha256-OI3N9zCKabDov2rZFzl8lJUXCcP7EmsGcGoP6DMXQCo="
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/MutationObserver.js/0.3.2/mutationobserver.min.js"
        integrity="sha256-BnmK1H6/rOiNcr4iGCjIyQqSO9hnMBZGJ0inXObQrTY="
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/fetch/2.0.3/fetch.min.js"
        integrity="sha256-aB35laj7IZhLTx58xw/Gm1EKOoJJKZt6RY+bH1ReHxs="
        crossorigin="anonymous"></script>
```

## Install with Yarn

```shell
# Install starting-blocks
yarn add starting-blocks
```

Before using *Starting Blocks* in your own project as a dependency you’ll need and to **create your own** `main.js` file
and your `ClassFactory.js` according to your website pages and blocks. Any other router dependencies can be
customize such as `TransitionFactory` to fit your own navigation needs.

*Starting Blocks* requires *jQuery*, *loglevel* and *jquery.waitforimages* as we do not provide it in our bundle. This is meant to allow CDN for
these libraries as you may need them in your project dependencies too!

You can use them for your *node_modules* and package them with your won project or use them as CDN like below:

```html
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"
        integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8="
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.waitforimages/2.2.0/jquery.waitforimages.min.js"
        integrity="sha256-b9bqxZdvRHQNAL/WJysGQ/mFHym7gGjEtruZ6zTNm7c="
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/loglevel/1.6.1/loglevel.min.js"
        integrity="sha256-AfynOKFjHWwlIEH3xY8lNK2sm2P2rxyiA749uLUXPh4="
        crossorigin="anonymous"></script>
```

**Be careful:** `jquery.waitforimages.min.js` library must be loaded as CDN only because it won’t be compatible with you *Webpack* build.

## Minimal usage

*Starting Blocks* is bundled with *NPM* in order to use *ES6* `import` syntax.
You won’t need to know where each class is stored, just use the *curly brace* syntax.
This bundle is already compiled in ES5, so you don’t need to setup Babel into your *node_modules* folder.

Minimal configuration:
```js
// Import Router class
import {
    Router,
    polyfills
} from 'starting-blocks'
import * as log from 'loglevel'

/*
 * Polyfills for Array.prototype.findIndex,  Array.prototype.find, Array.from, CustomEvent, 
 * console, requestAnimationFrame and cancelAnimationFrame
 */
polyfills()

/**
 * Config loglevel
 */
log.setLevel(0)

// Instanciate a Router
const routerOptions = {
    // ...
}
const router = new Router(routerOptions)

// Init router
router.init()
```

## A JS router made to work with HTML partial responses

This ES6 javascript router has been designed to handle as well complete HTML responses as
*partial* HTML responses to lighten backend process and bandwidth.
One of the most recurrent variable or member: `$cont` is always referring to the current page’ main-section.
This is the DOM section which is extracted at the end of each complete AJAX requests. When it detects that AJAX
response is *partial* it directly initialize `$cont` with the whole response data. Every new AJAX response will
be appended in the `#ajax-container` HTML section in order to smooth transitions between pages.

When `ajaxEnabled` is set to true and `window.fetch` is supported, **all links inside document body** are listened to load pages asynchronously and make transitions.

To declare a partial DOM section as the `$cont` you must add some classes and
data to your HTML tags.

```html
<div id="page-content-home"
     class="page-content page-content-ajax"
     data-node-type="page"
     data-node-name="home"
     data-is-home="1"
     data-meta-title="Home page">
</div>
```
- `id` *attribute* is obviously mandatory as it will be used to update your navigation and some other parts of your website. **Make sure that your ID is not equals to your `data-node-name`.**
- `page-content` *class* is essential in order to extract your DOM section during AJAX request. You can customize this class name in your `Router` options (`pageClass: "page-content"`).
- `data-node-type` *attribute* will be used to *route* your section to the corresponding JS class (in this example: page.js). **Every route class must extends the `AbstractPage` class**. Then you have to declare your routes in the `Router` construction (`'page' : Page`).
- `data-node-name` is used to name your page object **and to rename body ID after it.**
- `data-is-home`
- `data-meta-title` *attribute* will be used to change your new page *title* (`document.title`), it can be used in other cases such as some social network modules which require a clean page-title.

You’ll find `index.html` and `page1.html` examples files. You can even test them
by spawning a simple server with `npm run serve` command.
Then go to your favorite browser and type `http://localhost:8080`.

### Router dependencies

A Router needs:

- an options object in order to override default configuration with important parameters such as:
    - a `ClassFactory` object to link all `data-node-type` value to their *ES6* classes (you must import each class you’ll declare in your routes). You‘ll have to redefine a `ClassFactory` for each project you begin with *Starting Blocks*.
    - a `GraphicLoader` or extending class instance in order to trigger `show` or `hide` during AJAX requests.
    - a `TransitionFactory` object to link all `data-transition` value to their *ES6* classes.

You can look at the `example/src/app.js` file to see an instantiation example with few parameters.

### Router options

| Parameter | Type | Default | Description |
| --- | --- | --- | --- |
| ajaxEnabled | boolean | true | |
| pageClass | string | 'page-content' | |
| ajaxWrapperId | string | 'sb-wrapper' | |
| objectTypeAttr | string | 'data-node-type' | |
| noAjaxLinkClass | string | 'no-ajax-link' | |
| noPrefetchLinkClass | string | 'no-prefetch' | |
| pageBlockClass | string | '.page-block' | (with point) |
| lazyloadEnabled | boolean | false | |
| workerEnabled | boolean | false | |
| prefetchEnabled | boolean | true | |
| lazyloadSrcAttr | string | 'data-src' | |
| lazyloadClass | string | 'lazyload' | |
| lazyloadSrcSetAttr | string | 'data-srcset' | |
| lazyloadThreshold | number | 300 | Lazyload treshold |
| lazyloadThrottle | number | 150 | Duration of lazyload throttle |
| useCache | boolean | true | |
| classFactory | ClassFactory | ClassFactory | |
| graphicLoader | GraphicLoader | GraphicLoader | |
| transitionFactory | TransitionFactory | TransitionFactory | |

### Pages overridable methods

| Method | Description |
| --- | --- |
| `onLoad` | Called when all images and page are loaded. |
| `showEnded` | After show animation ended. |
| `onResize` | On viewport resize, this method is debounced. |
| `beforeLazyload` | Called before init lazyload images. |
| `onLazyImageSet` | After a lazyloaded image src switched. |
| `onLazyImageLoad` | After a lazyloaded image loaded. |
| `onLazyImageProcessed` | Before lazyload. |

### Block overridable methods

| Method | Description |
| --- | --- |
| `onLoad` | Called when all block images are loaded. |
| `onResize` | On viewport resize, this method is debounced. |
| `onPageReady` | Called once all page blocks have been created. |

### Events

| Const name | Event name | Description |
| --- | --- | --- |
| `BEFORE_PAGE_LOAD` | `SB_BEFORE_PAGE_LOAD` | Before Router initialize XHR request to load new page. |
| `AFTER_PAGE_LOAD` | `SB_AFTER_PAGE_LOAD` | After `window.fetch` XHR request succeeded. |
| `AFTER_DOM_APPENDED` | `SB_AFTER_DOM_APPENDED` | After Router appended new page DOM to `ajaxWrapperId`. |
| `AFTER_PAGE_BOOT` | `SB_AFTER_PAGE_BOOT` | After Router create new page instance. |
| `BEFORE_PAGE_SHOW` | `SB_BEFORE_PAGE_SHOW` | Before page begins to show, right after assets are loaded (images). |
| `AFTER_PAGE_SHOW` | `SB_AFTER_PAGE_SHOW` | After page showed. |
| `BEFORE_PAGE_HIDE` | `SB_BEFORE_PAGE_HIDE` | Before page begins to hide. *Be careful, this event must be triggered manually if hide() method is overriden.* |
| `AFTER_PAGE_HIDE` | `SB_AFTER_PAGE_HIDE` | After page hiding animation. *Be careful, this event must be triggered manually if hide() method is overriden.* |

⚠️ For those who use 2.1.7, we prefix events to avoid possible conflict with other libraries.

### Transitions

When AJAX navigation is enabled, transitions are used to manage animations between pages.

To manage transitions, you can set `data-transition` attribute with a name on each link.

```html
<a href="/contact" data-transition="fade">Contact</a>
```

Then, create a **TransitionFactory** class and pass it to the **Router** instance.
In this class, implement `getTransition (previousState, state, direction = null)` method.
This method is called on each transition and give you access to state informations :

- `previousState` and `state`
    - **transitionName** : `data-transition` attributes of the clicked link
    - **context** : equal to `"history"`, `"link"`

Example:
```javascript
// src/factories/TransitionFactory.js

import DefaultTransition from './transitions/DefaultTransition';
import FadeTransition from './transitions/FadeTransition';

export default class TransitionFactory {
    getTransition (previousState, state) {
        let transition = null

        switch (state.transitionName) {
            case 'fade':
                transition = new FadeTransition()
                break
            default:
                transition = new DefaultTransition()
                break
        }

        return transition
    }
}
```

To create a new transition you need to create a new class extending `AbstractTransition`. Implement `start()` method and use Promises to manages your animations.
⚠️ You have to wait for `this.newPageLoading` promise resolution to make sure the new page is accessible.
Then, you have access to old and new Page instances.

Example with fade animation:

```javascript
// src/transitions/FadeTransition.js

import AbstractTransition from '../AbstractTransition'

/**
 * Fade Transition class example. Fade Out / Fade In content.
 */
export default class FadeTransition extends AbstractTransition {
    /**
     * Entry point of the animation
     * Automatically called on init()
     */
    start () {
        // Wait new content and the end of fadeOut animation
        // this.newPageLoading is a Promise which is resolved when the new content is loaded
        Promise.all([this.newPageLoading, this.fadeOut()])
            // then fadeIn the new content
            .then(this.fadeIn.bind(this))
    }

    /**
     * Fade out the old content.
     * @returns {Promise}
     */
    fadeOut () {
        return new Promise((resolve) => {
            this.oldPage.$cont.animate({
                opacity: 0
            }, 400, 'swing', resolve)
        })
    }

    /**
     * Fade in the new content
     */
    fadeIn () {
        // Add display: none on the old container
        this.oldPage.$cont.hide()

        // Prepare new content css properties for the fade animation
        this.newPage.$cont.css({
            visibility: 'visible',
            opacity: 0
        })

        // fadeIn the new content container
        this.newPage.$cont.animate({ opacity: 1 }, 400, () => {
            document.body.scrollTop = 0
            // IMPORTANT: Call this method at the end
            this.done()
        })
    }
}
```

## Docs

To generate documentation, you’ll at least NodeJS v4.4 and to install ESDoc.

```bash
npm run doc;
```

Documentation will be available in `doc/` folder.

## Naming conventions

We drop jQuery in Starting-blocks v5 and change variables names and use some conventions.
We suffix every `DOMElement` variable with `Element`, `Container`, `Elements` or `Containers`.

Examples:
```js
let mainContainer = document.getElementById('main-container')
let imageContainer = document.getElementById('image-container')
let imageElements = imageContainer.querySelectorAll('.image')
```

## Improving Starting blocks

To work locally on *Starting blocks*, we provided some HTML example files.

- Install dependencies: `yarn`.
- Type `npm run dev` to improve Starting blocks locally.
- Type `npm run build` to optimize project in one file as: `main.js`.
- Type `npm run demo` to build demo project in `examples/` folder.

### Demo

To launch the example you need to change the `examples/srv/js/config/config.example.js` file with your own information.


CHANGELOG.md
# Changelog

## version 5.0.0 2018-03-06

- Drop jquery and waitForImages
- Remove `onLoad` method in `AbstractPage` and `AbstractBlock`

## Version 4.2.0: 2018-03-06

- Change `classFactory.getBlockInstance()` method parameters order (`page, $cont, nodeType`). Make sure to update your `classFactory` file
- You can now switch `classFactory.getBlockInstance()` to `async` and dynamically load your modules with `import('yourModule').then()`, see example for more details.
- Optional web worker feature to load pages content. Add `workerEnabled` router option to true to enable

## Version 4.1.0: 2018-02-20

- Change `classFactory.getPageInstance()` method parameters order and removed `isHome` parameter
- Added new `data` object parameter in history `add` method if you want to manually add specific data

## Version 4.0.0: 2018-02-08

- Big refactoring of all classes
- No more need to bind manually all **links** in Pages, Blocks or Nav, everything is handled by *Pjax* class
- Removed all Router constructor arguments except for `options` array
- `AbstractNav` has been removed, do not extend it anymore.
- Starting-blocks wrapper default ID is now `sb-wrapper`, you can change it in your `Router` options (i.e. `ajaxWrapperId: 'main-container'`)
- *jquery*, *loglevel* and *jquery.waitforimages* are declared external. You should provide them manually as file or as CDN in your project.
- *Pjax* is now using native `window.fetch` methods to perform XHR requests. Make sure to use necessary polyfills. **If window.fetch is not supported, ajax will be disabled.**


src/Router.js
/**
 * Copyright © 2016, Ambroise Maupate
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is furnished
 * to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 *
 * @file Router.js
 * @author Ambroise Maupate
 */
import $ from 'jquery'
import isMobile from 'ismobilejs'
import log from 'loglevel'
import Utils from './utils/Utils'
import State from './State'
import CacheProvider from './CacheProvider'
import Events from './Events'
import StatesStack from './StatesStack'
import TransitionFactory from './TransitionFactory'
import {
    BEFORE_PAGE_LOAD,
    AFTER_PAGE_LOAD,
    AFTER_DOM_APPENDED,
    AFTER_PAGE_BOOT
} from './EventTypes'

const DEFAULT_OPTIONS = {
    homeHasClass: false,
    ajaxEnabled: true,
    pageClass: 'page-content',
    objectTypeAttr: 'data-node-type',
    ajaxLinkTypeAttr: 'data-node-type',
    noAjaxLinkClass: 'no-ajax-link',
    navLinkClass: 'nav-link',
    activeClass: 'active',
    pageBlockClass: '.page-block',
    lazyloadEnabled: false,
    lazyloadSrcAttr: 'data-src',
    lazyloadClass: 'lazyload',
    lazyloadSrcSetAttr: 'data-srcset',
    lazyloadThreshold: 300,
    lazyloadThrottle: 150,
    $ajaxContainer: $('#ajax-container'),
    minLoadDuration: 0,
    preLoadPageDelay: 0,
    useCache: true
}

/**
 * Application main page router.
 */
export default class Router {
    /**
     * Create a new Router.
     *
     * ### Default options list:
     *
     * | Options | Default value |
     * | ----- | ----- |
     * | `homeHasClass` | `false` |
     * | `ajaxEnabled` | `true` |
     * | `pageClass` | "page-content" **without point!** |
     * | `objectTypeAttr` | "data-node-type" |
     * | `ajaxLinkTypeAttr`  | "data-node-type" |
     * | `noAjaxLinkClass` | "no-ajax-link" |
     * | `navLinkClass` | "nav-link" |
     * | `activeClass` | "active" |
     * | `useCache` | `true` |
     * | `pageBlockClass` | `".page-block"` **with point!** |
     * | `$ajaxContainer` | `$("#ajax-container")` |
     * | `lazyloadEnabled` | `false` |
     * | `lazyloadSrcAttr` | "data-src" |
     * | `lazyloadClass` | "lazyload" |
     * | `lazyloadSrcSetAttr` | "data-src-set" |
     * | `lazyloadThreshold` | `300` |
     * | `lazyloadThrottle` | `150` |
     * | `minLoadDuration` | `0` |
     * | `preLoadPageDelay` |  |
     *
     *
     * @param {Object} options
     * @param {ClassFactory} classFactory
     * @param {String} baseUrl
     * @param {GraphicLoader} loader
     * @param {AbstractNav} nav
     * @param {TransitionFactory} transitionFactory
     */
    constructor (options = {}, classFactory, baseUrl, loader, nav, transitionFactory = null) {
        /**
         * @type {Object}
         */
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options
        }

        /**
         * @type {TransitionFactory}
         */
        this.transitionFactory = null

        if (!baseUrl) {
            throw new Error('Router needs baseUrl to be defined.')
        }
        if (!loader) {
            throw new Error('Router needs a GraphicLoader instance to be defined.')
        }
        if (!classFactory) {
            throw new Error('Router needs a ClassFactory instance to be defined.')
        }
        if (!nav) {
            throw new Error('Router needs a Nav instance to be defined.')
        }
        if (this.options.ajaxEnabled) {
            this.transitionFactory = transitionFactory || new TransitionFactory()
        }

        /**
         * @type {ClassFactory}
         */
        this.classFactory = classFactory
        /**
         * @type {String}
         */
        this.baseUrl = baseUrl
        /**
         * @type {GraphicLoader}
         */
        this.loader = loader
        /**
         * @type {AbstractNav}
         */
        this.nav = nav
        this.nav.router = this
        /**
         * @type {State|null}
         */
        this.state = null
        /**
         * @type {Array}
         */
        this.formerPages = []
        /**
         * @type {null}
         */
        this.page = null
        this.transition = false
        this.loading = false
        this.statesStack = new StatesStack()
        this.$window = $(window)
        this.$body = $('body')

        this.deviceType = (isMobile.any === false) ? 'desktop' : 'mobile'
        Utils.addClass(this.$body[0], 'is-' + this.deviceType)

        /**
         * @deprecated use this.$window instead
         */
        this.window = this.$window
        this.currentRequest = null
        this.cacheProvider = new CacheProvider()
    }

    destroy () {
        if (this.options.ajaxEnabled) {
            window.removeEventListener('popstate', this.onPopState.bind(this), false)
        }
    }

    /**
     * Initialize Router events.
     */
    initEvents () {
        if (this.options.ajaxEnabled) {
            window.addEventListener('popstate', this.onPopState.bind(this), false)
            window.addEventListener(AFTER_PAGE_BOOT, this.trackGoogleAnalytics.bind(this), false)
        }
        /*
         * Init nav events
         */
        this.nav.initEvents(this)
    }
    /**
     * @private
     * @param  {Object} event
     * @return
     */
    onPopState (event) {
        if (typeof event.state !== 'undefined' && event.state !== null) {
            this.previousState = this.state
            this.direction = this.statesStack.getDirection(event.state)
            this.transition = true
            this.loadPage(event.state)
        }
    }

    /**
     * Booting need a jQuery handler for the jQuery container.
     *
     * Call this method in your `main.js` or app entry point **after** creating
     * the router and calling `initEvents` method.
     *
     * @param  {jQuery}  $cont The jQuery DOM element to boot in.
     * @param  {String}  context ["static" or custom string]
     * @param  {Boolean} isHome
     */
    boot ($cont, context, isHome) {
        if (context === 'static') {
            this.loadBeginDate = new Date()
        }

        const nodeType = $cont.attr(this.options.objectTypeAttr)
        this.page = this.classFactory.getPageInstance(nodeType, this, $cont, context, nodeType, isHome)

        /*
         * Replace current state
         * for first request
         */
        if (this.state === null) {
            this.state = new State(this)
            this.statesStack.push(this.state)
            window.history.replaceState(this.state, '', '')

            // Init first page
            this.pageLoaded()
        }

        if (context === 'ajax') {
            this.state.update(this.page)
        }

        Events.commit(AFTER_PAGE_BOOT, this.page)
    }

    pageLoaded () {
        const onShowEnded = this.page.onShowEnded.bind(this.page)
        this.loader.hide()

        Events.commit(AFTER_PAGE_LOAD, this.page.$cont)

        if (this.page.context === 'static') {
            this.page.show(onShowEnded)
        } else if (this.page.context === 'ajax') {
            // Update body id
            if (this.page.name !== null && this.page.name !== '') {
                document.body.id = this.page.name
                this.$body.addClass(this.page.name)
            }

            this.$body.addClass(this.page.type)
            this.page.show(onShowEnded)
        }
    }

    destroyPreviousPage () {
        // Hide formerPages - show
        if (this.page.context === 'ajax' && this.formerPages.length > 0) {
            const formerPage = this.formerPages[(this.formerPages.length - 1)]
            const formerPageDestroy = formerPage.destroy.bind(formerPage)
            /*
             * Very important,
             * DO NOT animate if there are more than 1 page
             * in destroy queue!
             */
            if (this.formerPages.length > 1) {
                formerPageDestroy()
            } else {
                formerPage.hide(formerPageDestroy)
            }
            this.formerPages.pop()
        }

        this.page.updateLazyload()
    }

    /**
     * @param e Event
     */
    onLinkClick (e) {
        const linkClassName = e.currentTarget.className
        const linkHref = e.currentTarget.href

        if (linkHref.indexOf('mailto:') === -1 &&
            linkClassName.indexOf(this.options.noAjaxLinkClass) === -1) {
            e.preventDefault()
            // Check if link is not active
            if (this.isNotCurrentPageLink(e.currentTarget)) {
                this.transition = true

                this.previousState = Object.assign({}, this.state)
                this.state = new State(this, e.currentTarget, {
                    previousType: this.page.type,
                    previousName: this.page.name,
                    navLinkClass: this.options.navLinkClass,
                    previousHref: window.location.href,
                    transitionName: $(e.currentTarget).data('transition')
                })
                this.statesStack.push(this.state)

                if (window.history.pushState) {
                    window.history.pushState(this.state, this.state.title, this.state.href)
                }
                this.loadPage(this.state)
            } else {
                log.debug('⛔️ Same page requested… do nothing.')
            }
        }
    }

    /**
     * Check if current target link is pointing to the same page resource.
     *
     * @param currentTarget
     * @return {boolean}
     */
    isNotCurrentPageLink (currentTarget) {
        const linkClassName = currentTarget.className
        return linkClassName.indexOf(this.options.activeClass) === -1 && !this.transition
    }

    /**
     * Perform a AJAX load for an History event.
     *
     * @param state
     * @private
     */
    loadPage (state) {
        if (this.currentRequest && this.currentRequest.readyState !== 4) {
            this.currentRequest.abort()
        }
        this.loader.show()
        this.loadBeginDate = new Date()

        Events.commit(BEFORE_PAGE_LOAD, state)

        this.transitionFactory.getTransition(this.previousState, state, this.direction)
            .init(this.page, this.doPageLoad(state))
            .then(() => {
                this.pageLoaded()
                this.destroyPreviousPage()
            })
    }

    /**
     * Actually load the state url resource.
     *
     * @param  {State} state
     * @private
     */
    doPageLoad (state) {
        return new Promise((resolve) => {
            if (this.options.useCache && this.cacheProvider.exists(state.href)) {
                log.debug('📎 Use cache-provider for: ' + state.href)
                resolve(this._onDataLoaded(this.cacheProvider.fetch(state.href), state))
            } else {
                this.currentRequest = $.ajax({
                    url: state.href,
                    dataType: 'html',
                    headers: {
                        // Send header to allow backends to
                        // send partial response for saving
                        // bandwidth and process time
                        'X-Allow-Partial': 1
                    },
                    // Need to disable cache to prevent
                    // browser to serve partial when no
                    // ajax context is defined.
                    cache: false,
                    type: 'get',
                    success: (data) => {
                        if (this.options.useCache) {
                            this.cacheProvider.save(state.href, data)
                        }
                        resolve(this._onDataLoaded(data, state))
                    }
                })
            }
        })
    }

    /**
     * Extract new page content from a complete HTML
     * page or a partial HTML response.
     *
     * @param data
     * @return {*}
     */
    extractPageContainer (data) {
        const $response = $($.parseHTML(data.trim()))
        if ($response.hasClass(this.options.pageClass)) {
            return $response
        } else {
            return $response.find('.' + this.options.pageClass)
        }
    }

    /**
     * @private
     * @param {Object} data jQuery AJAX response
     * @param {State} state
     */
    _onDataLoaded (data, state) {
        // Extract only to new page content
        // if the whole HTML is queried
        let $data = this.extractPageContainer(data)

        /*
         * Display data to DOM
         */
        $data.css('visibility', 'hidden')
        this.options.$ajaxContainer.append($data)

        Events.commit(AFTER_DOM_APPENDED, $data)

        /*
         * Push a copy object not to set it as null.
         */
        this.formerPages.push(this.page)

        // Init new page
        this.updatePageTitle($data)
        this.boot($data, 'ajax', state.isHome)

        return this.page
    }

    /**
     * Update page title against data-title attribute
     * from ajax loaded partial DOM.
     *
     * @param {jQuery} $data
     */
    updatePageTitle ($data) {
        if ($data.length && $data.attr('data-meta-title') !== '') {
            let metaTitle = $data.attr('data-meta-title')
            if (metaTitle !== null && metaTitle !== '') document.title = metaTitle
        }
    }

    /**
     * Send a GA page view event when context is AJAX.
     */
    trackGoogleAnalytics (event) {
        if (event && event.detail.context === 'ajax') {
            if (typeof ga !== 'undefined') {
                log.debug('🚩 Push Analytics for: ' + window.location.pathname)
                window.ga('send', 'pageview', {'page': window.location.pathname, 'title': document.title})
            }
        }
    }
}
